atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/CqlClientFactory.java

```diff
@@ -0,0 +1,168 @@
+                .withCredentials(config.credentials().username(), config.credentials().password())
+                .withCompression(ProtocolOptions.Compression.LZ4)
+                .withLoadBalancingPolicy(loadBalancingPolicy(config, servers))
+                .withPoolingOptions(poolingOptions(config))
+                .withQueryOptions(queryOptions(config))
+                .withRetryPolicy(DefaultRetryPolicy.INSTANCE)
+                .withoutJMXReporting()
+                .withThreadingOptions(new ThreadingOptions());
+
+        proxy.ifPresent(proxyAddress -> clusterBuilder.withNettyOptions(new SocksProxyNettyOptions(proxyAddress)));
+
+        sslOptions(config).ifPresent(clusterBuilder::withSSL);
+
+        Cluster cluster = clusterBuilder.build();
+
+        ThreadFactory threadFactory = new ThreadFactoryBuilder()
+                .setNameFormat(cluster.getClusterName() + "-session" + "-%d")
+                .build();
+
+        return CqlClientImpl.create(cluster, Executors.newCachedThreadPool(threadFactory), initializeAsync);
+    }
+
+    private static Optional<RemoteEndpointAwareJdkSSLOptions> sslOptions(CassandraKeyValueServiceConfig config) {
+        if (config.sslConfiguration().isPresent()) {
+            SSLContext sslContext = SslSocketFactories.createSslContext(config.sslConfiguration().get());
+            return Optional.of(RemoteEndpointAwareJdkSSLOptions.builder()
+                    .withSSLContext(sslContext)
+                    .build());
+        } else if (config.ssl().orElse(false)) {
+            return Optional.of(RemoteEndpointAwareJdkSSLOptions.builder().build());
+        }
+        return Optional.empty();
+    }
+
+    private static PoolingOptions poolingOptions(CassandraKeyValueServiceConfig config) {
+        return new PoolingOptions()
+                .setMaxConnectionsPerHost(HostDistance.LOCAL, config.poolSize())
+                .setMaxConnectionsPerHost(HostDistance.REMOTE, config.poolSize())
+                .setPoolTimeoutMillis(config.cqlPoolTimeoutMillis());
+    }
+
+    private static QueryOptions queryOptions(CassandraKeyValueServiceConfig config) {
+        return new QueryOptions().setFetchSize(config.fetchBatchCount());
+    }
+
+    private static LoadBalancingPolicy loadBalancingPolicy(CassandraKeyValueServiceConfig config,
+            Set<InetSocketAddress> servers) {
+        // Refuse to talk to nodes twice as (latency-wise) slow as the best one, over a timescale of 100ms,
+        // and every 10s try to re-evaluate ignored nodes performance by giving them queries again.
+        // Note we are being purposely datacenter-irreverent here, instead relying on latency alone
+        // to approximate what DCAwareRR would do;
+        // this is because DCs for Atlas are always quite latency-close and should be used this way,
+        // not as if we have some cross-country backup DC.
+        LoadBalancingPolicy policy = LatencyAwarePolicy.builder(new RoundRobinPolicy()).build();
```

Comment:
This is nice vis-a-vis the DC aware one, because you don't need to know your location. We are trying out implementing DC/rack awareness in CassandraService and it is not the simplest thing to figure out your node's location.

Reply:
Non actionable.